### PR TITLE
[Feat] Binary focal loss - fixing weighting factor (alpha, 1-alpha)

### PR DESCRIPTION
**Pull request by xen0f0n**
Sorry for the delay! 
Finally, after months  #506 #493 I'm submitting the fix for the focal loss for the binary task. In the [original paper](https://arxiv.org/abs/1708.02002) only the weighting factor alpha for the rare class is given and it's not clearly stated whether the weight for the other class is ```1-alpha``` or something else entirely. 

I'm going with ```(alpha, 1-alpha)```.

### CODE REVIEW

File: test/test_losses.py [hunk 1]
```diff
@@ -9,6 +9,71 @@
 from torch.testing import assert_allclose
 
 
+class TestFocalLossWithLogits:
+    def test_smoke_none(self, device):
```
Comment (edgarriba) [inline test/test_losses.py:13]: can you add `dtype` in the signature as well ?
`def test_smoke_none(self, device, dtype):`
and then create the input tensor with that type

File: test/test_losses.py [hunk 2]
```diff
@@ -9,6 +9,71 @@
...
+        num_classes = 1
+        logits = torch.rand(2, num_classes, 3, 2).to('cpu')
```
Comment (edgarriba) [inline test/test_losses.py:15]: why you send to cpu ? do this: `logits = torch.rand(2, num_classes, 3, 2, device=device, dtype=dtype)`

File: test/test_losses.py [hunk 3]
```diff
@@ -9,6 +9,71 @@
...
+        labels = labels.to(device).long()
+        assert (
+            kornia.losses.binary_focal_loss_with_logits(
+                logits, labels, alpha=0.5, gamma=2.0, reduction="mean"
+            ).shape == ()
+        )
+
+    # TODO: implement me
+    def test_jit(self, device):
+        pass
```
Comment (edgarriba) [inline test/test_losses.py:60]: could you try to implement this one too ? see an example here: https://github.com/kornia/kornia/blob/master/test/geometry/test_linalg.py#L82 (omit lines 83-85)

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: test/test_losses.py [hunk 4]
```diff
@@ -9,6 +9,79 @@
...
+        assert (
+            kornia.losses.binary_focal_loss_with_logits(
+                logits, labels, alpha=0.5, gamma=2.0, reduction="mean"
+            ).shape == ()
+        )
+
+    def test_jit(self, device, dtype):
+        @torch.jit.script
+        def op_script(logits, labels):
+            return kornia.losses.binary_focal_loss_with_logits(logits, labels, alpha=0.5, gamma=2.0, reduction="none")
```
Comment (edgarriba) [inline test/test_losses.py:58]: this local function is not needed

File: kornia/losses/focal.py [hunk 5]
```diff
@@ -122,3 +122,93 @@ def forward(  # type: ignore
...
+        inp: torch.Tensor,
+        target: torch.Tensor,
+        alpha: float = .25,
+        gamma: float = 2.0,
+        reduction: str = 'none',
+        eps: float = 1e-8) -> torch.Tensor:
+    r"""Function that computes Binary Focal loss.
+
+    """
+    if not torch.is_tensor(inp):
```
Comment (edgarriba) [inline kornia/losses/focal.py:137]: @xen0f0n you need to modify this line by: `if not isinstance(inp, torch.Tensor):` since `torch.is_tensor(inp)` is not supported by JIT

**Comment by edgarriba** [comment]
@xen0f0n we are releasing in a couple of weeks kornia v0.5. Would be great to have this in the release

**Comment by edgarriba** [comment]
@xen0f0n any update on this ?

**Comment by xen0f0n** [comment]
@edgarriba No updates, sorry... I haven't picked it up for more than a month. I'll work on it during the weekend and get back to you, if that's ok with you.

**Comment by edgarriba** [comment]
@xen0f0n make sure you do proper rebase

**Comment by xen0f0n** [comment]
@edgarriba the failed test is not PR-related, right? it fails on some docs folder...

File: kornia/losses/focal.py [hunk 6]
```diff
@@ -122,3 +122,93 @@ def forward(  # type: ignore
...
+        [1] https://arxiv.org/abs/1708.02002
+    """
+
+    def __init__(self, alpha: float, gamma: float = 2.0,
+                 reduction: str = 'none') -> None:
+        super(BinaryFocalLossWithLogits, self).__init__()
+        self.alpha: float = alpha
+        self.gamma: float = gamma
+        self.reduction: str = reduction
+        self.eps: float = 1e-6
```
Comment (edgarriba) [inline kornia/losses/focal.py:208]: just noticed that eps is not exposed to the module constructor

File: test/test_losses.py [hunk 7]
```diff
@@ -9,6 +9,75 @@
...
+        logits = torch.rand(2, num_classes, 3, 2).to(device)
+        labels = torch.rand(2, 1, 3, 2) * num_classes
+        labels = labels.to(device).long()
+
+        logits = utils.tensor_to_gradcheck_var(logits)  # to var
+        assert gradcheck(
+            kornia.losses.binary_focal_loss_with_logits,
+            (logits, labels, alpha, gamma),
+            raise_exception=True,
+        )
```
Comment (edgarriba) [inline test/test_losses.py:78]: add a small test comparing the results from the functional and the module

File: kornia/losses/focal.py [hunk 8]
```diff
@@ -122,3 +122,93 @@ def forward(  # type: ignore
...
+def binary_focal_loss_with_logits(
+        inp: torch.Tensor,
+        target: torch.Tensor,
+        alpha: float = .25,
+        gamma: float = 2.0,
+        reduction: str = 'none',
+        eps: float = 1e-8) -> torch.Tensor:
+    r"""Function that computes Binary Focal loss.
+
+    """
```
Comment (edgarriba) [inline kornia/losses/focal.py:136]: it's worth adding the documentation here too, including a very small example. Check e.g the functionalities from kornia.color

**Review by edgarriba — COMMENTED** [review]
LGTM. Just very quick additions. The failing test might be a random error from the read the docs server side.

File: kornia/losses/focal.py [hunk 9]
```diff
@@ -122,3 +122,93 @@ def forward(  # type: ignore
...
+    else:
+        raise NotImplementedError("Invalid reduction mode: {}"
+                                  .format(reduction))
+    return loss
+
+
+class BinaryFocalLossWithLogits(nn.Module):
+    r"""Criterion that computes Focal loss.
+
+    According to [1], the Focal loss is computed as follows:
```
Comment (edgarriba) [inline kornia/losses/focal.py:175]: I think the docs error comes from the reference here. We support bibtex style for the bibliography. Check this:
https://github.com/kornia/kornia/blob/master/kornia/feature/hardnet.py#L17
https://github.com/kornia/kornia/blob/master/docs/source/references.bib

File: kornia/losses/focal.py (hunk 6 above)
Comment (xen0f0n) [reply kornia/losses/focal.py:203]: Should it be exposed? You mean as a kwarg?

File: kornia/losses/focal.py (hunk 8 above)
Comment (xen0f0n) [reply kornia/losses/focal.py:134]: Looking at the dice loss, I noticed that there is an "Examples" section at the module (and not at functional) so I added it there as well, just to be consistent.

**Comment by edgarriba** [comment]
@xen0f0n one question: did you do a git rebase anytime here ? I'm just noticing that the circleci tests are the old ones and not include the recent with TPU support.

**Comment by xen0f0n** [comment]
@edgarriba I picked it up from yesterday's rebase. Did I mess this up again?

**Comment by xen0f0n** [comment]
@edgarriba yes, of course I messed rebase! I've got it this time!

File: test/test_losses.py (hunk 7 above)
Comment (xen0f0n) [reply test/test_losses.py:78]: Something like this? 

    def test_same_output(self, device, dtype):
        num_classes = 1
        logits = torch.rand(2, num_classes, 3, 2, dtype=dtype, device=device)
        labels = torch.rand(2, 3, 2, dtype=dtype, device=device)

        kwargs = {"alpha": 0.25, "gamma": 2.0, "reduction": 'mean'}

        assert kornia.losses.binary_focal_loss_with_logits(logits, labels, **kwargs) == \
               kornia.losses.BinaryFocalLossWithLogits(**kwargs)(logits, labels)

**Comment by edgarriba** [comment]
@xen0f0n yeah, seems that the rebase went somehow wrong. Check for instance https://github.com/kornia/kornia/pull/784 and you'll see that the CircleCI tests are different. I suggest to do as follows: `git pull --rebase upstream master` - solve conflicts and you might have to force push.

**Comment by xen0f0n** [comment]
@edgarriba This can't be right...  +28,304 −4,835 
I think I've messed this really bad...

**Comment by edgarriba** [comment]
@xen0f0n something went wrong again with rebase. What about a fresh commit ?

**Comment by xen0f0n** [comment]
@edgarriba  I rebased to kornia/master, but at this point I'm not sure what's wrong.. 
Would it make sense to re-fork the repo and PR just the files focal.py and test_losses.py?

**Comment by edgarriba** [comment]
@xen0f0n yes, that's what I meant for fresh commit :)
